src/lib.rs

```diff
@@ -381,14 +383,18 @@ pub fn encode_to_fmt<T: AsRef<[u5]>>(
     hrp: &str,
     data: T,
 ) -> Result<fmt::Result, Error> {
-    check_hrp(&hrp)?;
-    match Bech32Writer::new(hrp, fmt) {
+    let hrp_lower = match check_hrp(&hrp)? {
+        Case::Upper => hrp.to_lowercase(),
+        Case::Lower | Case::None => String::from(hrp),
```

Comment:
The string allocation could be avoided by using `Cow<str>`. 

```suggestion-2
        Case::Upper => Cow::Owned(hrp.to_lowercase()),
        Case::Lower | Case::None => Cow::Borrowed(hrp),
```

Reply:
Great point. I'll make that modification.

Reply:
Added in d2e4c49.